### DISCUSSION TITLE
Control how much query result items being update at once? Hard cap? Paginated results?

### DISCUSSION THREAD

**Discussion by pc223** [Flow-Launcher/Flow.Launcher#553] [Ideas]
I notice when **first time** launch Flow and make a query (especially Plugin.Program), the **UI** of Flow hang a bit, like 2-5 seconds. Probably because of too many query results? (in my testing, the query results count is ~5000 results, related to this https://github.com/Flow-Launcher/Flow.Launcher/issues/551)

I think it doesn't really have much value to update to UI thousands of results at once.

Maybe a hard cap at number of need-to-update-to-UI-results? Or even better, paginated results?

Paginated results similar to REST API, with requests for next and prev page, results count, seems a bit overkill, but more controlled.
<img src="https://user-images.githubusercontent.com/10551242/124384518-cda55a80-dcfb-11eb-9f31-46d7a54f5aa8.png" width=400>

**Comment by taooceros** [comment]
I don't think Flow will hang for 5000 results in UI update because the use of lazy load image and listview virtualization (which means only the results that is viewable will be updated). From my test, it remains active even with 100000 results (iterating all files from Program Files).

However, it is possible that the program plugin is consuming too much resources because it is executing parallel.

**Reply by pc223** [reply]
Do you experience the hang when first time launch Flow and query the Program plugin?

**Reply by taooceros** [reply]
Just for checking, is the hang meaning that Flow is not responsible, or simply the results are not popped out but the loading bar is loading?

**Reply by taooceros** [reply]
> Do you experience the hang when first time launch Flow and query the Program plugin?

Let me restart and take a try.

**Reply by pc223** [reply]
Flow is not responsible, like can't type, no loading show 😅

**Reply by pc223** [reply]
I put some logging in, seems that the query done really quick around ~10ms, but the UI hang around 2-5 seconds. So it's likely related to the UI part

**Reply by taooceros** [reply]
Are you in debug mode or release mode?

**Reply by pc223** [reply]
I build in release config

**Reply by pc223** [reply]
It's only hang at the first time launch Flow tho, after that it's fine, maybe cache related?

**Reply by pc223** [reply]
I just downloaded a CI artifact, still able to reproduce that hang

**Reply by taooceros** [reply]
> It's only hang at the first time launch Flow tho, after that it's fine, maybe cache related?

What if program plugin is disabled?

**Reply by pc223** [reply]
Seems I can only make a long hang with Program plugin, other plugins not clear

**Reply by pc223** [reply]
It's pretty weird, if I open Flow.Launcher.exe, and query a plugin (ControlPanel) before query the Program plugin, it's fine. So the first time Flow query any plugin, seems to "clear" the "block" somehow.

**Reply by taooceros** [reply]
Maybe I know what's happening. It has happened a long time ago when I drafting #195.

I will reach out soon since I am doing something else currently.

**Reply by pc223** [reply]
> Maybe I know what's happening. It has happened a long time ago when I drafting #195.
> 
> I will reach out soon since I am doing something else currently.

Sure thing!

**Reply by jjw24** [reply]
hmm i cant replicate with starting up and search for something from programs plugin.

dont limit result though that's a bad design, paginated is fine as long as it is obvious to the user how they can move to the next set and is not slow. but best to find out the root cause.

**Comment by taooceros** [comment]
I think I failed to reply yesterday. Check out build from #554

**Reply by pc223** [reply]
Just quick test that PR, I can't reproduce the hang anymore in normal use, but it still hangs when I run debug with release config in IDE Rider. Not sure exactly why yet, maybe related to how the debugger works with how many items being updated at once. The hang maybe related to high RAM too.

<img src="https://user-images.githubusercontent.com/10551242/124434767-76a49180-dd9e-11eb-9d1a-c9a5b33a4bca.gif" width=400>

**Reply by taooceros** [reply]
Is it possible that this is because the program plugin is unable to load cache so it takes a while to index all programs?

**Reply by taooceros** [reply]
Does Flow remain responsive when the results haven't popped out?

**Reply by pc223** [reply]
When it hangs, the status in Task Manager is "Not responding" and the mouse icon is rotating like below, indicate something block the main thread UI I think. Not sure tho.

![image](https://user-images.githubusercontent.com/10551242/124603928-a1711180-de94-11eb-8acd-a86869350865.png)

I don't think it's the index, all the timing I can measure return back really short (~10ms), all the update results seem done really quickly, I highly suspect the UI event handler of changing results or something like that, not the query code. I suspect the hang happens after this line, not before.
```
OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
```

Can you give me some testing strategies? Like logging positions or debugging tools, I'm planning to try to build on Visual Studio later.

EDIT:

Confirmed,I set a breakpoint and the hang happens after this line (in `BulkAddAll()`, not before)

```
OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
```

![image](https://user-images.githubusercontent.com/10551242/124607303-e9456800-de97-11eb-8e17-79c58c69881b.png)

**Reply by jjw24** [reply]
can you give your pc specs please, because i am not able to repro- i am on 16 ram, latest windows 10 update, 4 cores

**Reply by jjw24** [reply]
also does it happen with actually an installed version of flow not via ide?

**Reply by taooceros** [reply]
Interesting. I do remember that the bulk update will cause some delay at the first time rendering, which leads me to make small amount of update using normal update method. However, that won't work well for large amount of data changed.

**Reply by pc223** [reply]
> can you give your pc specs please, because i am not able to repro- i am on 16 ram, latest windows 10 update, 4 cores

I'm on a pretty powerful PC, 16ram, 10400 intel, Win10-21H1, but when I use Flow, I tend to open a lot of programs, RAM usage tends to be around 13GB. So I'm still a bit suspect it's related to RAM usage.

> also does it happen with actually an installed version of flow not via ide?

Without taooceros's PR #554 , it happens at both, with and without IDE. With that PR, it only happens with IDE in debug mode (release config).

> Interesting. I do remember that the bulk update will cause some delay at the first time rendering, which leads me to make small amount of update using normal update method. However, that won't work well for large amount of data changed.

Yeah, so that means the PR #554 is not a root cause solution right? @taooceros But can you reproduce this in your machine?
